File: app/src/main/java/com/example/githubrepo/BaseSearchActivity.java
/*
 * Copyright (c) 2016 Razeware LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.example.githubrepo;

import android.content.SharedPreferences;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

import com.example.githubrepo.models.Repository;
import com.example.githubrepo.services.BusProvider;
import com.example.githubrepo.services.GitHubService;
import com.example.githubrepo.services.event.BusEvent;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.Comparator;
import java.util.List;

import javax.inject.Inject;

import butterknife.ButterKnife;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by sarahneo on 20/2/17.
 */

public abstract class BaseSearchActivity extends AppCompatActivity {

    protected Drawable x;

    //protected GitHubService mService;
    protected SharedPreferences mSharedPref;

    @Inject
    BusProvider busProvider;
    @Inject
    GitHubService gitHubService;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        ((MyApplication) getApplication()).inject(this);

        mSharedPref = PreferenceManager.getDefaultSharedPreferences(this);
    }

    @Override
    protected void onStart() {
        super.onStart();
        try {
            busProvider.register(this);
        } catch (IllegalArgumentException e) {}
    }

    @Override
    protected void onStop() {
        super.onStop();
        busProvider.unregister(this);
    }

    public void post(Object event) {
        busProvider.post(event);
    }

    public Object getEvent(BusEvent.EventType type) {
        return busProvider.getEvent(type);
    }

}

File: app/src/main/java/com/appcessible/gitsearcher/BaseSearchActivity.kt
package com.appcessible.gitsearcher

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.appcessible.gitsearcher.databinding.ActivityMainBinding
import com.appcessible.gitsearcher.services.BusProvider
import com.appcessible.gitsearcher.services.GitHubService
import com.appcessible.gitsearcher.services.event.BusEvent
import javax.inject.Inject

abstract class BaseSearchActivity : AppCompatActivity() {

    protected lateinit var sharedPref: SharedPreferences
    protected lateinit var binding: ActivityMainBinding
    @Inject internal lateinit var busProvider: BusProvider
    @Inject internal lateinit var service: GitHubService

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater, null, false)
        setContentView(binding.root)
        (application as MyApplication).inject(this)
        sharedPref = getPreferences(Context.MODE_PRIVATE)
    }

    override fun onStart() {
        super.onStart()
        try {
            busProvider.register(this)
        } catch (e: IllegalArgumentException) {
            Log.e(BaseSearchActivity::class.java.simpleName, "Unable to register busProvider", e)
        }
    }

    override fun onStop() {
        super.onStop()
        busProvider.unregister(this)
    }

    fun post(event: Any?) {
        event?.let { busProvider.post(it) }
    }

    fun getEvent(type: BusEvent.EventType) = busProvider.getEvent(type)
}